pallets/worker/src/lib.rs

```diff
@@ -397,29 +400,73 @@ decl_module! {
                 epoch_info.set(&epoch);
                 // regeneate cache and store the dag
                 let l_dag = DAG::new(header.number.into());
-                cache = l_dag.cache.clone();
+                cache = l_dag.cache;
             }
 
             let mut stream = RlpStream::new();
             Self::rlp_append(header.clone(), &mut stream);
             let rlp_header: Vec<u8> = stream.out();
 
             let signer = Signer::<T, T::AuthorityId>::any_account();
-
+            let dataset_info = StorageValueRef::persistent(b"light-client-worker::ethash-dataset");
+            let dataset = match dataset_info.get::<Vec<u8>>() {
+                Some(dataset) => Some(dataset.unwrap()),
+                None => {
+                    // FIXME(shekohex): move this code to a function guarded by std feature.
+                    // so we can only generate the dataset on std supported targets.
+                    debug::native::info!("Generate dataset for stored epoch: {}", stored_epoch);
+                    // lock the storage.
+                    let mut lock = StorageLock::<Time>::new(b"light-client-worker::ethash-dataset");
+                    let maybe_dataset = if let Ok(_guard) = lock.try_lock() {
+                        let dataset_size = ethash::get_full_size(stored_epoch as usize);
+                        let mut dataset = Vec::with_capacity(dataset_size);
+                        dataset.resize(dataset_size, 0u8);
+                        ethash::make_dataset(&mut dataset, &cache);
+                        dataset_info.set(&dataset);
+                        Some(dataset)
+                    } else {
+                        // no work to be done here, until the dataset generation is complete.
+                        None
+                    };
+                    maybe_dataset
+                },
+            };
+            let dataset = match dataset {
+                Some(dataset) => dataset,
+                // this skips the current invocation, as there is another worker is generating the
+                // dataset.
+                None => return
+            };
+            let should_generate_dataset = should_generate_dataset(U256::from(header.number), stored_epoch);
+            debug::native::info!("Should we generate the dataset? {}", should_generate_dataset);
+            if should_generate_dataset {
+                // FIXME(shekohex): lock the storage here to avoid regeneation.
```

Comment:
Currently thinking is to maintain 2 datasets at once using differently named storage entries as well as a bit as to which one is current.
- `first-dataset` and `second-dataset` return datasets
- `is-first-curr` should return a boolean for whether the dataset stored at `first-dataset` is the dataset currently intended to be processed for generating proofs.

Lets say we start from no datasets..
- We first generate a dataset in `first-dataset` location.
- When regeneration threshold is met, we generate second dataset in `second-dataset` location.
- When an epoch is changing, we check `is-first-curr`. If yes, we set it to false, if no, we set it to true.
- When `is-first-curr == false` and we hit regeneration threshold, we simply overwrite storage in `first-dataset`.

Reply:
Got it, starting working on that!